src/org/daisy/dotify/formatter/impl/row/SegmentProcessor.java

```diff
@@ -760,59 +849,460 @@ int getForceCount() {
         return forceCount;
     }
 
-    @Override
-    public boolean isEmpty() {
-        return empty;
+    /**
+     * Returns true if there is an active (not completed) row in the
+     * processor.
+     *
+     * @return true if there is an active row
+     */
+    private boolean hasCurrentRow() {
+        return currentRow != null;
     }
 
-    @Override
-    public boolean hasCurrentRow() {
-        return currentRow != null;
+    /**
+     * Starts a new active row. Note that if {@link #hasCurrentRow()} returns
+     * true, {@link #flushCurrentRow()} should be called first.
+     *
+     * @param left  the left margin
+     * @param right the right margin
+     */
+    private void newCurrentRow(MarginProperties left, MarginProperties right) {
+        currentRow = processorContext.getRowDataProps().configureNewEmptyRowBuilder(left, right);
     }
 
-    @Override
-    public int getUnusedLeft() {
-        return minLeft;
+    /**
+     * Adds to the running force count value.
+     *
+     * @param value the value to add
+     */
+    private void addToForceCount(double value) {
+        forceCount += value;
     }
 
-    @Override
-    public int getUnusedRight() {
-        return minRight;
+    /**
+     * Returns true if the block processor has a pending list item.
+     *
+     * @return true if the block processor has a pending list item, false otherwise
+     */
+    private boolean hasListItem() {
+        return item != null;
     }
 
-    @Override
-    public void newCurrentRow(MarginProperties left, MarginProperties right) {
-        currentRow = spc.getRdp().configureNewEmptyRowBuilder(left, right);
+    /**
+     * Discards the pending list item.
+     */
+    private void discardListItem() {
+        item = null;
     }
 
-    @Override
-    public Builder getCurrentRow() {
-        return currentRow;
+    /**
+     * Gets the pending list item, or null if {@link #hasListItem()} returns false.
+     *
+     * @return returns the pending list item, or null
+     */
+    private ListItem getListItem() {
+        Objects.requireNonNull(item);
+        return item;
     }
 
-    @Override
-    public void addToForceCount(double value) {
-        forceCount += value;
+    /* ========================= */
+    /*       Inner classes       */
+    /* ========================= */
+
+    private interface CurrentResult {
```

Comment:
Why put everything in one file?
I would prefer to have a separate file for this interface and a separate file for every implementation, instead of this super-large file

Reply:
AFAIK you can't put inner classes in a separate file, that's the problem.

If you look at what these classes do, it's obvious that they only make sense in the context of `SegmentProcessor`. I think it's because Joel has an aversion towards nested classes that the code ended up being the way it was, with these weird constructs that made it hard to understand.

That this makes the file a bit bigger is a drawback but I think it's worse to separate code over multiple classes when they are not actually separable. I think there are other ways to better modularize this code.